File: codecheck/codecheck/url_in_documentation.py
from __future__ import annotations

from re import Pattern, compile as re

from astroid.nodes import ClassDef
from pylint.typing import TYPE_CHECKING

from rulebook_pylint.checkers.rulebook_checkers import RulebookChecker

if TYPE_CHECKING:
    from pylint.lint import PyLinter


class UrlInDocumentation(RulebookChecker):
    name: str = 'url-in-documentation'
    msgs = {
        'W9901': ('Missing URL in docstring.', 'missing-url', 'Add URL.'),
        'W9902': ('Invalid URL in docstring.', 'invalid-url', 'Fix URL.'),
    }

    _REGEX: Pattern = re(r'([a-z])([A-Z])')

    def visit_classdef(self, node: ClassDef) -> None:
        name = node.name
        if name.startswith('Rulebook') or \
            name.startswith('Base') or \
            name.startswith('Test'):
            return
        if name.endswith('Rule'):
            base_name = name[:-4]
        elif name.endswith('Checker'):
            base_name = name[:-7]
        else:
            return

        doc = node.doc_node.value if node.doc_node else None
        if not doc:
            self.add_message('missing-url', node=node)
            return

        kebab_name = self._REGEX.sub(r'\1-\2', base_name).lower()
        if doc == f'See detail: https://hanggrian.github.io/rulebook/rules/#{kebab_name}':
            return
        self.add_message('invalid-url', node=node)


def register(linter: PyLinter) -> None:
    linter.register_checker(UrlInDocumentation(linter))

File: codecheck/src/url-in-documentation.ts
import { Rule } from 'eslint';
import { ClassDeclaration } from 'estree';

export default {
    create(context: Rule.RuleContext) {
        return {
            ClassDeclaration(node: ClassDeclaration) {
                if (!node.id || node.id.name.startsWith('Rulebook')) {
                    return;
                }

                let baseName = node.id.name;
                if (baseName.endsWith('Rule')) {
                    baseName = baseName.slice(0, -4);
                } else if (baseName.endsWith('Checker')) {
                    baseName = baseName.slice(0, -7);
                } else {
                    return;
                }

                const sourceCode = context.sourceCode;
                const jsdoc =
                    sourceCode
                        .getCommentsBefore(node)
                        .find(comment => comment.type === 'Block' && comment.value.startsWith('*'));

                if (!jsdoc) {
                    context.report({ node, message: 'Missing URL.' });
                    return;
                }
                const kebabName = baseName.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
                if (jsdoc
                    .value
                    .includes(
                        '{@link ' +
                        `https://hanggrian.github.io/rulebook/rules/#${kebabName}|See detail` +
                        '}',
                    )
                ) {
                    return;
                }
                context.report({ node, message: 'Invalid URL.' });
            },
        };
    },
};
